server/src/main/java/org/opensearch/search/GenericSearchExtBuilder.java

```diff
@@ -0,0 +1,160 @@
+        switch (valueType) {
+            case SIMPLE:
+                genericObj = in.readGenericValue();
+                break;
+            case MAP:
+                genericObj = in.readMap();
+                break;
+            case LIST:
+                genericObj = in.readList(r -> r.readGenericValue());
+                break;
+            default:
+                throw new IllegalStateException("Unable to construct GenericSearchExtBuilder from incoming stream.");
+        }
+    }
+
+    public static GenericSearchExtBuilder fromXContent(XContentParser parser) throws IOException {
+        // Look at the parser's next token.
+        // If it's START_OBJECT, parse as map, if it's START_ARRAY, parse as list, else
+        // parse as simpleVal
+        XContentParser.Token token = parser.currentToken();
+        ValueType valueType;
+        Object genericObj;
+        if (token == XContentParser.Token.START_OBJECT) {
+            genericObj = parser.map();
+            valueType = ValueType.MAP;
+        } else if (token == XContentParser.Token.START_ARRAY) {
+            genericObj = parser.list();
+            valueType = ValueType.LIST;
+        } else if (token.isValue()) {
+            genericObj = parser.objectText(); // .objectBytes() ??
+            valueType = ValueType.SIMPLE;
+        } else {
+            throw new XContentParseException("Unknown token: " + token);
+        }
+
+        return new GenericSearchExtBuilder(genericObj, valueType);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return EXT_BUILDER_NAME.getPreferredName();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeInt(valueType.getValue());
+        switch (valueType) {
+            case SIMPLE:
+                out.writeGenericValue(genericObj);
+                break;
+            case MAP:
+                out.writeMap((Map<String, Object>) genericObj);
+                break;
+            case LIST:
+                out.writeCollection((List<Object>) genericObj, StreamOutput::writeGenericValue);
+                break;
+            default:
+                throw new IllegalStateException("Unknown valueType: " + valueType);
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        switch (valueType) {
+            case SIMPLE:
+                return builder.field(EXT_BUILDER_NAME.getPreferredName(), genericObj);
+            case MAP:
+                return builder.field(EXT_BUILDER_NAME.getPreferredName(), (Map<String, Object>) genericObj);
+            case LIST:
+                return builder.field(EXT_BUILDER_NAME.getPreferredName(), (List<Object>) genericObj);
+            default:
+                return null;
+        }
+    }
+
+    // We need this for the equals method.
+    Object getValue() {
+        return genericObj;
+    }
```

Comment:
If a client receives one of these `GenericExtBuilder` instances, should we let them access the underlying object?

Reply:
We talked about this over Slack:

Michael Froh 1:59 PM
I guess my only remaining question is what we expect a client to do if they get one of these GenericExtBuilders in their deserialized search response (assuming they’re using the REST high-level client).

As written, they should probably treat it as an opaque box, and not care about that part of the response. That’s probably fine — at least the client doesn’t break. If they want, they can use the `toXContent` method to dump it to JSON and do something with that.

Austin Lee 2:02 PM
Yes, I think it's really there as a last resort?  They should grab the correct ext builder if they want to do something useful.

Michael Froh 2:06 PM
Sounds good to me :+1:

Reply:
I intentionally made it not public (but accessible to unit tests).  I don't think there's a use case for allowing clients to access the underlying objects.

Reply:
I think it make sense: if user care about them, they should be configured, it not - we just absorb the payload.